sdk/servicebus/Azure.Messaging.ServiceBus/src/Core/RetriableContext.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.ServiceBus.Amqp;
+using Azure.Messaging.ServiceBus.Diagnostics;
+using Microsoft.Azure.Amqp;
+
+namespace Azure.Messaging.ServiceBus.Core
+{
+    internal class RetriableContext
```

Comment:
On the fence of whether I'm a fan of this type of abstraction or feel that it is just adding a level of indirection and overhead to key networking operations for a very small gain in reuse....

Reply:
I think it is important that we have some abstraction for executing service operations. Otherwise we end up needing to carefully place the same 20-30 lines of code around the body of each service method. In terms of the implementation of the abstraction, there may be a better approach out there. @pakrym for his thoughts

Reply:
@jsquire do you inline retry code into call-sites in EH?

Reply:
The responsibility for determining whether or not to retry is delegated to the policy;  the exception management is delegated to an transport exception class;    The paths in Event Hubs implement their own wrapping while loop that call those two things and respect their decisions.

I considered doing something similar to this where I had the while loop in a class that took an action to perform, but the logic is minimal, there were only 3 or 4 call sites, and each was at a low level where I preferred not to have the overhead of indirection.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/9921